### ISSUE TITLE
Saving/reading hypsography with InputOutput is not fully automated

### ISSUE THREAD

**Issue by Sbozzolo** [CliMA/ClimaCore.jl#1603]
Right now, the `InputOutput` module has to know about the details of the hypsography to save/read it from file. Adding a new adaption requires adjusting this module. It should be possible to automate this.

**Comment by charleskawczynski** [comment]
Isn't this kind of true with every data structure? I'm not actually sure why we don't use a recursive and generic pattern here.

**Comment by Sbozzolo** [comment]
> Isn't this kind of true with every data structure? I'm not actually sure why we don't use a recursive and generic pattern here.

Yes, exactly. At least for hypsography, we should be able to do everything in a generic way

**Comment by charleskawczynski** [comment]
I've thought about this, and I remembered that there is a tension between two things:

 - Having a fully generic (or nearly fully generic) way to read/write data
 - Maintaining forward / backwards compatibility (i.e., brittleness)

Similar to serialization, a fully generic / automated solution will mean that any changes to data structures will result in backwards-incompatibility. We could try to reduce this by effectively adding patches to "standardize" outputs for certain data structures, but those patches will grow over time.

I've used fully generic solutions in my past, and they're nice because they require little to no maintenance. But you basically cannot restart simulations between different versions.
